### PR TITLE
store: fix marshaling with sync.Pool

### PR DESCRIPTION
**Pull request by GiedriusS**
Do not forget to check the length of a slice returned by sync.Pool.
Annotate the whole function with comments to aid understanding of it.

Cover the case with a test that fails on `main`.

Fixes https://github.com/thanos-io/thanos/issues/4591.

cc @yeya24 as you've noticed this too

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

### CODE REVIEW

**Review by yeya24 — APPROVED** [review]
LGTM. Thanks for the quick fix.

File: pkg/store/storepb/custom.go [hunk 1]
```diff
@@ -497,32 +497,45 @@ func (m *SeriesResponse) Marshal() (dAtA []byte, err error) {
...
-		if m.respBuf == nil {
-			poolBuf := m.respPool.Get()
-			if poolBuf == nil {
-				respBuf = make([]byte, size)
-				m.respBuf = &respBuf
-			} else {
-				m.respBuf = poolBuf.(*[]byte)
-				respBuf = *m.respBuf
-			}
+		m.respBuf = nil
```
Comment (yeya24) [inline pkg/store/storepb/custom.go:503]: Is this line necessary?

Comment (bwplotka) [reply pkg/store/storepb/custom.go:503]: Yea, I don't think so

File: pkg/store/storepb/custom.go [hunk 2]
```diff
@@ -497,32 +497,45 @@ func (m *SeriesResponse) Marshal() (dAtA []byte, err error) {
...
+
+		n, err := m.MarshalToSizedBuffer(respBuf)
+		if err != nil {
+			return nil, err
+		}
+		return respBuf[len(respBuf)-n:], nil
+	}
+
+	// Fast path with sync.Pool.
+	// m.respBuf must not be nil so that it would be returned to the pool.
```
Comment (bwplotka) [inline pkg/store/storepb/custom.go:513]: Does it need to non nil if we do the work below? Can we always do that and remove any outside call pool Get?

File: pkg/store/storepb/custom.go [hunk 3]
```diff
@@ -497,32 +497,45 @@ func (m *SeriesResponse) Marshal() (dAtA []byte, err error) {
...
+	// Last sanity check of the size before the marshaling.
+	if cap(respBuf) < size {
+		if m.respPool != nil {
+			m.respPool.Put(&respBuf)
+		}
+		respBuf = make([]byte, size)
 	}
+	m.respBuf = &respBuf
 
+	// Possibly trim it so that there wouldn't be left-over "garbage" in the slice.
```
Comment (bwplotka) [inline pkg/store/storepb/custom.go:538]: Let's add TODO, but I think we should not trim on 100% calls as between small and large response we would have similar allocs and latency and if we would not pool

File: pkg/store/storepb/custom_test.go [hunk 4]
```diff
@@ -525,3 +526,47 @@ func TestMatchersToString_Translate(t *testing.T) {
 
 	}
 }
+
+// Tests whether the Marshal() function properly checks the size
+// of a slice returned from a sync.Pool.
+// Reference: https://github.com/thanos-io/thanos/issues/4591.
```
Comment (bwplotka) [inline pkg/store/storepb/custom_test.go:532]:
```suggestion
// Regression test against https://github.com/thanos-io/thanos/issues/4591.
```

**Review by bwplotka — APPROVED** [review]
Amazing, LGTM, just one nit.

File: pkg/store/storepb/custom.go (hunk 2 above)
Comment (GiedriusS) [reply pkg/store/storepb/custom.go:513]: I'm not sure I understood this comment. The callers using the old function pass `nil` thus we need to handle this case - it is not guaranteed that the pool exists

Comment (bwplotka) [reply pkg/store/storepb/custom.go:512]: I know, i thought we can simplify, and allow this method to `Get`, never pass it from outside - not a biggie

**Review by bwplotka — COMMENTED** [review]
Looks like we still have this data race we discussed here here https://cloud-native.slack.com/archives/CL25937SP/p1629797444067700

**Review by bwplotka — APPROVED** [review]
It does not fix everything, but fix some of things. LGTM and let's iterate.

**Pull request merged**